src/commands/key2ds.rs

```diff
@@ -0,0 +1,166 @@
+            };
+
+            let class = record.class();
+            let ttl = record.ttl();
+            let owner = record.owner();
+
+            // Of the records that we see, we only care about DNSKEY records
+            let ScannedRecordData::Dnskey(dnskey) = record.data() else {
+                continue;
+            };
+
+            // if ignore_sep is specified, we accept any key
+            // otherwise, we only want SEP keys
+            if !self.ignore_sep && !dnskey.is_secure_entry_point() {
+                continue;
+            }
+
+            let key_tag = dnskey.key_tag();
+            let sec_alg = dnskey.algorithm();
+            let digest_alg = self.determine_hash(sec_alg);
+
+            if digest_alg == DigestAlg::GOST {
+                return Err("Error: the GOST algorithm is deprecated and must not be used. Try a different algorithm.".into());
+            }
+
+            let digest = dnskey
+                .digest(&owner, digest_alg)
+                .map_err(|e| format!("Error computing digest: {e}"))?;
+
+            let ds = Ds::new(key_tag, sec_alg, digest_alg, digest).expect(
+                "Infallible because the digest won't be too long since it's a valid digest",
+            );
+
+            let rr = Record::new(owner, class, ttl, ds);
+
+            if self.write_to_stdout {
+                println!("{}", rr.display_zonefile(false));
+            } else {
+                let owner = owner.fmt_with_dot();
+                let sec_alg = sec_alg.to_int();
+                let filename = format!("K{owner}+{sec_alg:03}+{key_tag:05}.ds");
+                let mut out_file = File::create(&filename)
+                    .map_err(|e| format!("Could not create file \"{filename}\": {e}"))?;
+                writeln!(out_file, "{}", rr.display_zonefile(false))
+                    .map_err(|e| format!("Could not write to file \"{filename}\": {e}"))?;
+
+                // This is different from ldns, but I think writing out the
+                // filename we wrote to is useful:
+                println!("Wrote DS record to: {filename}");
+            }
+        }
+
+        Ok(())
+    }
+
+    fn determine_hash(&self, sec_alg: SecAlg) -> DigestAlg {
+        // If a specific algorithm was set, use that
+        if self.one {
+            return DigestAlg::SHA1;
+        } else if self.two {
+            return DigestAlg::SHA256;
+        } else if self.four {
+            return DigestAlg::SHA384;
+        } else if let Some(a) = self.algorithm {
+            return a;
+        }
+
+        // Otherwise we try to determine a similar hash to the key
+        match sec_alg {
+            SecAlg::RSASHA256
+            | SecAlg::RSASHA512
+            | SecAlg::ED25519
+            | SecAlg::ED448
+            | SecAlg::ECDSAP256SHA256 => DigestAlg::SHA256,
+            SecAlg::ECDSAP384SHA384 => DigestAlg::SHA384,
+            SecAlg::ECC_GOST => DigestAlg::GOST,
+            _ => DigestAlg::SHA1,
```

Comment:
Is using SHA-1 as the fallback a good idea? If new algorithms were to get introduced and we forget to update the above list, they would fallback to using SHA-1, right?

Reply:
It's the current ldns behaviour: https://github.com/NLnetLabs/ldns/blob/b87b8e1d212b43f225891570ebc5045c6331dd8a/examples/ldns-key2ds.c#L48C1-L48C14

If we want to change it, let's take it to a separate PR, where we can discuss this in depth.

Reply:
Ok